system/core/tags.php

```diff
@@ -0,0 +1,184 @@
+			$html.= '<div id="selectedTags" style="display:inline-block;">';
+			if (!empty($selectedTags)) foreach ($selectedTags as $id => $tag) {
+				$html.= '<div id="tagSel_'.$id.'" class="tag">';
+				$html.= '<span title="'.$tag['description'].'">'.$tag['label'].'</span>';
+				if (isUser()) $html.= ' <span style="color:#8f0000; cursor:pointer;" title="'.__('remove-tag').'" onclick="deselectTag(\''.$id.'\');">⨯</span>';
+				$html.= '</div>';
+			}
+			$html.= '</div>';
+			if ((isUser() && !empty($deselectedTags)) || isAdmin()) {
+				$html.= '<select id="remainingTags" style="padding:0px;" onchange="selectTag(this.value);">';
+				$html.= '<option value="_add_" selected="selected" disabled="disabled">'.__('add').'...</option>';
+				if (!empty($deselectedTags)) foreach ($deselectedTags as $id => $tag) {
+					$html.= '<option id="tagRem_'.$id.'" value="'.$id.'">'.$tag['label'].'</option>';
+				}
+				if (isAdmin()) $html.= '<option value="">'.__('new').'...</option>';
+				$html.= '</select>';
+			}
+		}
+		return $html;
+	}
+	
+	/*
+		Function: hypha_indexTags
+		returns list of tags for the given page
+
+		Parameters:
+		$pageListNode - DOMElement containing page settings
+		$lang - language for tag index
+	*/
+	function hypha_indexTags($pageListNode, $lang) {
+		global $hyphaHtml;
+		
+		if (!$pageListNode || !$lang) return '';
+		
+		$language = hypha_pageGetLanguage($pageListNode, $lang);
+		$url = $language->getAttribute('id').'/'.$language->getAttribute('name');
+		
+		ob_start();
+?>
+<script>
+	function selectTag(id) {
+		if (!id) addTag();
+		else {
+			hypha('<?=$url?>', 'tagSelect', id, function(response) {document.getElementById('tagList').innerHTML = response;});
+			document.getElementById('tagList').innerHTML = 'loading...';
+		}
+	}
+	function deselectTag(id) {
+		hypha('<?=$url?>', 'tagDeselect', id, function(response) {document.getElementById('tagList').innerHTML = response;});
+		document.getElementById('tagList').innerHTML = 'loading...';
+	}
+	function addTag() {
+		var name = prompt('<?=__('enter-label')?>', '');
+		if (name) {
+			var description = prompt('<?=__('enter-description')?>', '');
+			if (description) {
+				var json = '{ "name" : "' + name + '", "description" : "' + description + '"}';
+				hypha('<?=$url?>', 'tagNew', json, function(response) {document.getElementById('tagList').innerHTML = response;});
+				document.getElementById('tagList').innerHTML = 'loading...';
+			}
+		}
+	}
+</script>
+<?php
+		$hyphaHtml->writeScript(ob_get_clean());
+		
+		return tagList($pageListNode, $lang);
+	}
+	
+	registerCommandCallback('tagSelect', 'tagSelectCallback');
+	registerCommandCallback('tagDeselect', 'tagDeselectCallback');
+	registerCommandCallback('tagNew', 'tagNewCallback');
+	
+	function tagSelectCallback($id) {
+		global $hyphaXml;
+		global $O_O;
+		
+		$lang = $O_O->getContentLanguage();
+		
+		$hyphaXml->lockAndReload();
+		$pageListNode = hypha_getPage($lang, $O_O->getRequest()->getPageName());
```

Comment:
I'm not so sure I like this, since it partially redoes request routing (url to page). The existince of `getPageName()` makes this a lot better than it would be otherwise, though (and I'm not sure if there's a better and also easy way for this...).

Reply:
We just discussed this and the newtext refactoring will probably significantly change the way commands are handled anyway, so we should probably just leave this as it is now and include this in the newtext design instead.